Manhattan problem for the most part solved

MilanG-git opened the pull request:
Hi everyone!
This is my solution to the Manhattan problem.
The  issue I found and couldn't find a workaround yet is with the largest file. Namely with:
`python3 MilanG-git-Manhattan.py Manhattan-testHVD2.in -d -t`
Although, the right score is printed, the path can't be printed because the maximum recursion depth is reached and the program aborted.
Please let me know where to improve.
Best regards,
Milan

Lorenzo360 commented:
Hi Milan, 
your can change the maximum recursion limit with `sys.setrecursionlimit(numberOfRecursions)`
Best, 
Lorenz

MilanG-git commented:
Thanks, that did the trick. It can now process the largest file.
Cheers,
Milan

robnic2718 commented:
Dear @MilanG-git 

well done. Great code structure and readability and good variable naming. I ran your code and except for one thing all works properly. :)  

Here are my suggestions for (possible) improvements:

*1.* You read in the files assuming weight matrices do not follow directly after another, which is why there is an error message when one tries to run your code for the input-file "Manhattan-testHV2.in". You could therefore implement some kind of counter to let your program know when the next weight matrix starts. 

*2.* In order to be able to also have negative weights (I know this is not specified in the task, but still) you could change the stopping criteria in line 124 ... i.e. instead of checking whether the value is 0 check whether you reached the index position [0,0]. If negative weights are present 0 could also occur (especially for integers) in the middle of the matrix, which would break the program

*3.* Your code for creating the dynamic programming matrix is efficient, however your function route() slows down your program. My suggestion would be that you implement the backtracking such that you avoid deleting the rows/columns of the dynamic programming matrix. Recursion + iterative access and deleting of memory is quite time-consuming. I guess it will not be too much work to change that, but the improvement will be significant.

Cheers Nicolas

MilanG-git commented:
> Dear @MilanG-git
> 
> well done. Great code structure and readability and good variable naming. I ran your code and except for one thing all works properly. :)
> 
> Here are my suggestions for (possible) improvements:
> 
> _1._ You read in the files assuming weight matrices do not follow directly after another, which is why there is an error message when one tries to run your code for the input-file "Manhattan-testHV2.in" (matrices following directly after another). You could therefore implement some kind of counter to let your program know when the next weight matrix starts.
> 
> _2._ In order to be able to also have negative weights (I know this is not specified in the task, but still) you could change the stopping criteria in line 124 ... i.e. instead of checking whether the value is 0 check whether you reached the index position [0,0]. If negative weights are present 0 could also occur (especially for integers) in the middle of the matrix, which would break the functionality.
> 
> _3._ Your code for creating the dynamic programming matrix is efficient, however your function route() slows down your program. My suggestion would be that you implement the backtracking such that you avoid deleting the rows/columns of the dynamic programming matrix. Recursion + iterative access and deleting of memory is quite time-consuming. Apart from that np.delete returns a copy (to avoid that you could use inplace=True), which means you are doing many many unnecessary copies of large matrices. I guess it will not be too much work to change that, but the improvement will be significant.
> 
> Cheers Nicolas

Hi @robnic2718 ,

Thanks for your review, it was really helpful.

ad 1.: When reading the description of the task I thought that the resulting manhattan matrix should be a square (n x n), but that's not the case in "Manhattan-testHV2.in". That was my error in reasoning.

ad 2.: True and changed.

ad 3.: I was wondering how to speed up the process and this was it. I changed it so that the backtracking uses the indexes instead of deleting rows/columns.. and the increase in speed is immense.

Cheers;
Milan

erdogabormate commented:
Hi Milan,

I got to review the final version of your code, so my review will add less value than the previous ones, but you got some good feedback so I am not preocupied.

I find it elegant that you look for float/int very early in the code, and spare yourself output formatting hassle at the end, and store data in the optimal format along the code. I like your coding style, and argparse is something I will start using in the next assignment (if there will be arguments required) as it makes your code more concise.
Your code length does not substantially differ from other codes I have seen, but your runtime for the biggest matrix is excellent, about 3 seconds.
Also your route function is the first one I see that is easy to read and just made me understand what I am doing wrong (I am still working on a good implementation). So unfortunately no improvement feedback to give, but I guess giving only positive feedback after many improvements already made by you is not that wrong.
